File: perfwatch/python/tests/test_mock_collector.py
import sys
from types import ModuleType

import pytest

from perfwatch.collectors.mock import MockCollector, get_mock_snapshot
from perfwatch.collectors.native import NativeCollector


def test_mock_snapshot_preserves_baseline_and_evolves() -> None:
    baseline = get_mock_snapshot(0)
    next_snapshot = get_mock_snapshot(1)

    assert baseline["timestamp_ms"] == 1_710_000_000_000
    assert baseline["cpu"]["usage_percent"] == 42.5
    assert baseline["top_processes"][0]["name"] == "mock_process"
    assert next_snapshot["timestamp_ms"] == 1_710_000_001_000
    assert next_snapshot["cpu"]["usage_percent"] > baseline["cpu"]["usage_percent"]


def test_mock_collectors_advance_independently() -> None:
    first_collector = MockCollector()
    second_collector = MockCollector()

    assert first_collector.collect() == second_collector.collect()
    assert first_collector.collect()["timestamp_ms"] == 1_710_000_001_000
    assert second_collector.collect()["timestamp_ms"] == 1_710_000_001_000


def test_mock_snapshot_rejects_negative_index() -> None:
    with pytest.raises(ValueError, match="sample_index must be non-negative"):
        get_mock_snapshot(-1)


def _fake_native_module(calls: list[int]) -> ModuleType:
    module = ModuleType("perfwatch_native")

    class FakeWindowsCollector:
        def __init__(self) -> None:
            self.sample_index = 0

        def collect(self) -> dict[str, object]:
            calls.append(self.sample_index)
            snapshot = get_mock_snapshot(self.sample_index)
            self.sample_index += 1
            return snapshot

    setattr(module, "WindowsCollector", FakeWindowsCollector)
    return module


def test_native_wrapper_uses_stateful_windows_collector(monkeypatch: pytest.MonkeyPatch) -> None:
    calls: list[int] = []
    monkeypatch.setitem(sys.modules, "perfwatch_native", _fake_native_module(calls))

    first_collector = NativeCollector()
    second_collector = NativeCollector()

    assert first_collector.collect()["timestamp_ms"] == 1_710_000_000_000
    assert first_collector.collect()["timestamp_ms"] == 1_710_000_001_000
    assert second_collector.collect()["timestamp_ms"] == 1_710_000_000_000
    assert calls == [0, 1, 0]


def test_native_collector_never_falls_back_to_mock(monkeypatch: pytest.MonkeyPatch) -> None:
    monkeypatch.setitem(sys.modules, "perfwatch_native", None)
    collector = NativeCollector()

    with pytest.raises(RuntimeError, match="live Windows collector unavailable"):
        collector.collect()

File: perfwatch/cpp/tests/test_mock_collector.cpp
#include <cassert>

#include "perfwatch/collector.hpp"

void test_mock_snapshot_shape();
void test_proc_stat_parser();
void test_meminfo_parser();
void test_proc_pid_stat_parser();
void test_power_supply_parser();

int main() {
    test_mock_snapshot_shape();
    test_proc_stat_parser();
    test_meminfo_parser();
    test_proc_pid_stat_parser();
    test_power_supply_parser();

    const auto baseline = perfwatch::make_mock_snapshot(0);
    assert(baseline.timestamp_ms == 1'710'000'000'000);
    assert(baseline.cpu.usage_percent == 42.5);
    assert(baseline.memory.used_bytes == 17'179'869'184ULL);
    assert(baseline.battery.percent == 78.0);
    assert(baseline.top_processes.at(0).estimated_power_score == 0.42);

    const auto indexed = perfwatch::make_mock_snapshot(3);
    assert(indexed.timestamp_ms == 1'710'000'003'000);
    assert(indexed.cpu.usage_percent == 47.0);
    assert(indexed.memory.used_bytes == 17'381'195'776ULL);
    assert(indexed.battery.percent == 77.25);
    assert(indexed.top_processes.at(0).rss_bytes == 281'018'368ULL);

    const auto repeated = perfwatch::make_mock_snapshot(3);
    assert(indexed.timestamp_ms == repeated.timestamp_ms);
    assert(indexed.cpu.usage_percent == repeated.cpu.usage_percent);
    assert(indexed.memory.used_bytes == repeated.memory.used_bytes);
    assert(indexed.battery.percent == repeated.battery.percent);
    assert(indexed.top_processes.at(0).estimated_power_score ==
           repeated.top_processes.at(0).estimated_power_score);

    const auto peak = perfwatch::make_mock_snapshot(10);
    const auto descending = perfwatch::make_mock_snapshot(11);
    assert(peak.cpu.usage_percent == 57.5);
    assert(descending.cpu.usage_percent == 56.0);
    assert(peak.cpu.usage_percent > descending.cpu.usage_percent);

    const auto last_discharging = perfwatch::make_mock_snapshot(40);
    const auto first_charging = perfwatch::make_mock_snapshot(41);
    const auto next_cycle = perfwatch::make_mock_snapshot(80);
    assert(!last_discharging.battery.charging);
    assert(last_discharging.battery.percent == 68.0);
    assert(first_charging.battery.charging);
    assert(first_charging.battery.percent == 68.25);
    assert(!next_cycle.battery.charging);
    assert(next_cycle.battery.percent == 78.0);

    perfwatch::MockCollector first_collector;
    perfwatch::MockCollector second_collector;
    assert(first_collector.collect().timestamp_ms == 1'710'000'000'000);
    assert(first_collector.collect().timestamp_ms == 1'710'000'001'000);
    assert(second_collector.collect().timestamp_ms == 1'710'000'000'000);

    return 0;
}
